dtls.c

```diff
@@ -2848,22 +2883,47 @@ dtls_send_client_hello(dtls_context_t *ctx, dtls_peer_t *peer,
     p += cookie_length;
   }
 
+  for (index = 0; handshake->cipher_suites[index] != TLS_NULL_WITH_NULL_NULL; ++index) {
+    dtls_cipher_t code = handshake->cipher_suites[index];
+    if (known_and_supported_cipher_suite(ctx, code, 1)) {
+        cipher_suites_size += sizeof(uint16);
+#ifdef DTLS_ECC
+        ecdsa = ecdsa || is_tls_ecdhe_ecdsa_with_aes_128_ccm_x(code);
+#endif /* DTLS_ECC */
+    }
+    // ignore not supported cipher-suite
+    // credentials callback is missing
+  }
+
+  if (cipher_suites_size == 0) {
+    dtls_crit("no supported cipher-suite provided!\n");
+    return dtls_alert_fatal_create(DTLS_ALERT_HANDSHAKE_FAILURE);
+  }
+
   /* add known cipher(s) */
-  dtls_int_to_uint16(p, cipher_size - 2);
+  dtls_int_to_uint16(p, cipher_suites_size);
   p += sizeof(uint16);
 
-  if (ecdsa) {
-    dtls_int_to_uint16(p, TLS_ECDHE_ECDSA_WITH_AES_128_CCM_8);
-    p += sizeof(uint16);
-    dtls_int_to_uint16(p, TLS_ECDHE_ECDSA_WITH_AES_128_CCM);
-    p += sizeof(uint16);
+  for (index = 0; handshake->cipher_suites[index] != TLS_NULL_WITH_NULL_NULL; ++index) {
```

Comment:
As this loop is almost the same as the one twenty lines above, it might be worth spending some thought if this could be combined. The first thing that comes to mind here is something like this (untested).
[combine-offer-loops.patch.txt](https://github.com/eclipse/tinydtls/files/8761619/combine-offer-loops.patch.txt)

Reply:
I currently test an alternative design, keeping the p of the size field and set the value later. I test it and on success I will push it.

Reply:
```
  /* keep pointer to size of cipher suites */
  p_cipher_suites_size = p;
  /* skip size of cipher suites field */
  p += sizeof(uint16);

  /* add known cipher(s) */
  for (index = 0; handshake->cipher_suites[index] != TLS_NULL_WITH_NULL_NULL; ++index) {
    dtls_cipher_t code = handshake->cipher_suites[index];
    if (known_and_supported_cipher_suite(ctx, code, 1)) {
      dtls_int_to_uint16(p, code);
      p += sizeof(uint16);
#ifdef DTLS_ECC
      ecdsa = ecdsa || is_tls_ecdhe_ecdsa_with_aes_128_ccm_x(code);
#endif /* DTLS_ECC */
    }
    /* ignore not supported cipher-suite
       credentials callback is missing */
  }

  cipher_suites_size = (p - p_cipher_suites_size) - sizeof(uint16);
  if (cipher_suites_size == 0) {
    dtls_crit("no supported cipher-suite provided!\n");
    return dtls_alert_fatal_create(DTLS_ALERT_HANDSHAKE_FAILURE);
  }

  /* set size of known cipher suites */
  dtls_int_to_uint16(p_cipher_suites_size, cipher_suites_size);

```

Reply:
The tests have failed. It requires a fix ;-).